src/ray/object_manager/object_manager.cc

```diff
@@ -349,17 +350,132 @@ ray::Status ObjectManager::SendObjectData(const ObjectID &object_id,
   return status;
 }
 
-ray::Status ObjectManager::Wait(const std::vector<ObjectID> &object_ids,
-                                uint64_t timeout_ms, int num_ready_objects,
+ray::Status ObjectManager::Wait(const std::vector<ObjectID> &object_ids, int64_t wait_ms,
+                                uint64_t num_required_objects, bool wait_local,
                                 const WaitCallback &callback) {
-  // TODO: Implement wait.
+  UniqueID wait_id = UniqueID::from_random();
+
+  if (wait_local) {
+    return ray::Status::NotImplemented("Wait for local objects is not yet implemented.");
+  }
+
+  RAY_CHECK(wait_ms >= 0);
+  RAY_CHECK(num_required_objects != 0);
+  RAY_CHECK(num_required_objects <= object_ids.size());
+  if (object_ids.size() == 0) {
+    callback(std::vector<ObjectID>(), std::vector<ObjectID>());
+  }
+
+  // Initialize fields.
+  active_wait_requests_.emplace(wait_id, WaitState(*main_service_, wait_ms, callback));
+  auto &wait_state = active_wait_requests_.find(wait_id)->second;
+  wait_state.object_id_order = object_ids;
+  wait_state.wait_ms = wait_ms;
+  wait_state.num_required_objects = num_required_objects;
+  for (auto &oid : object_ids) {
+    if (local_objects_.count(oid) > 0) {
+      wait_state.found.insert(oid);
+    } else {
+      wait_state.remaining.insert(oid);
+    }
+  }
+
+  if (wait_state.remaining.empty()) {
+    WaitComplete(wait_id);
+  } else {
+    for (auto &oid : wait_state.remaining) {
+      // Lookup remaining objects.
+      wait_state.requested_objects.insert(oid);
+      RAY_CHECK_OK(object_directory_->LookupLocations(
+          oid, [this, wait_id](const std::vector<ClientID> &client_ids,
+                               const ObjectID &object_id) {
+            auto &wait_state = active_wait_requests_.find(wait_id)->second;
+            if (!client_ids.empty()) {
+              wait_state.remaining.erase(object_id);
+              wait_state.found.insert(object_id);
+            }
+            wait_state.requested_objects.erase(object_id);
+            if (wait_state.requested_objects.empty()) {
+              AllWaitLookupsComplete(wait_id);
+            }
+          }));
+    }
+  }
   return ray::Status::OK();
 }
 
+void ObjectManager::AllWaitLookupsComplete(const UniqueID &wait_id) {
+  auto &wait_state = active_wait_requests_.find(wait_id)->second;
+  if (wait_state.found.size() >= wait_state.num_required_objects ||
+      wait_state.wait_ms == 0) {
+    // Requirements already satisfied.
+    WaitComplete(wait_id);
+  } else {
+    for (auto &oid : wait_state.remaining) {
+      // Subscribe to object notifications.
+      wait_state.requested_objects.insert(oid);
+      RAY_CHECK_OK(object_directory_->SubscribeObjectLocations(
+          wait_id, oid, [this, wait_id](const std::vector<ClientID> &client_ids,
+                                        const ObjectID &object_id) {
+            auto &wait_state = active_wait_requests_.find(wait_id)->second;
```

Comment:
There is an issue with the way this code is currently structured because of the fact that the callback registered in `SubscribeObjectLocations` may now get called directly. It's possible that right now we will not actually see a bug, but that is only because of the specific ordering of calls made on the object directory by the object manager, and that seems quite brittle. Here is a scenario where I think the code would break:

1. `SubscribeObjectLocations` gets called on objects A and B (e.g., for a `Pull`, or for a different `Wait`). Locations for both are cached in the object directory.
2. `Wait` is called on objects A and B, with 1 object required.
3. **In the same call stack**, `SubscribeObjectLocations` is called on object A. The cached locations are found, this callback fires, and the wait request completes and is erased from `active_wait_requests_`.
4. Again, **in the same call stack**, `SubscribeObjectLocations` is called on object B. `active_wait_requests_.find(wait_id)` will fail (silently).